src/solver/src/steady_mesh_adaptive_solver.C

```diff
@@ -93,71 +95,93 @@ namespace GRINS
 
     for ( unsigned int r_step = 0; r_step < _mesh_adaptivity_options.max_refinement_steps(); r_step++ )
       {
-        std::cout << "==========================================================" << std::endl
+        std::cout << std::endl
+                  << "==========================================================" << std::endl
                   << "Adaptive Refinement Step " << r_step << std::endl
                   << "==========================================================" << std::endl;
 
-        // Solve the forward problem
-        context.system->solve();
+        converged = this->amr_helper(context,r_step,true);
+        if (converged)
+          break;
 
-        if( context.output_vis )
-          {
-            context.postprocessing->update_quantities( *(context.equation_system) );
-            context.vis->output( context.equation_system );
-          }
+      } // r_step for-loop
 
-        // Solve adjoint system
-        if(context.do_adjoint_solve)
-          this->steady_adjoint_solve(context);
+    // this will print output the error estimates and/or meshes after the final AMR run
+    if (_mesh_adaptivity_options.max_refinement_steps() > 0)
+      if (! converged)
+        {
+          std::cout << std::endl
+                    << "==========================================================" << std::endl
+                    << "Post-AMR Output " << std::endl
+                    << "==========================================================" << std::endl;
+          this->amr_helper(context,_mesh_adaptivity_options.max_refinement_steps(),false);
```

Comment:
Just checking that I'm reading this correclty: this will ensure we do one final solve after the last refinement step? This was actually missing before, we did an AMR step, but didn't solve to update those new dofs, they were just constrained from the previous solution (which is not entirely egregious, but not as accurate).

Reply:
Right: previously a 1-step AMR run wouldn't actually refine the mesh at all, just calculate the errors and output the original mesh if requested. Intuitively, I would assume that a 10-step AMR run would refine the mesh 10 times and give me 11 error estimates and QoI values (from the original mesh and after the 10 refinements)

Reply:
I do need to make a small modification here so that I can get my QoI error estimates without doing a refinement